// balena-cli/build/commands/env/rm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const command_2 = require("../../command");
const ec = require("../../utils/env-common");
const lazy_1 = require("../../utils/lazy");
const validation_1 = require("../../utils/validation");
class EnvRmCmd extends command_2.default {
    async run() {
        const { args: params, flags: opt } = this.parse(EnvRmCmd);
        await command_2.default.checkLoggedIn();
        const { confirm } = await Promise.resolve().then(() => require('../../utils/patterns'));
        await confirm(opt.yes || false, 'Are you sure you want to delete the environment variable?', undefined, true);
        const balena = lazy_1.getBalenaSdk();
        await balena.pine.delete({
            resource: ec.getVarResourceName(opt.config, opt.device, opt.service),
            id: params.id,
        });
    }
}
exports.default = EnvRmCmd;
EnvRmCmd.description = lazy_1.stripIndent `
		Remove a config or env var from an application, device or service.

		Remove a configuration or environment variable from an application, device
		or service, as selected by command-line options.

		${ec.rmRenameHelp.split('\n').join('\n\t\t')}

		Interactive confirmation is normally asked before the variable is deleted.
		The --yes option disables this behavior.
`;
EnvRmCmd.examples = [
    '$ balena env rm 123123',
    '$ balena env rm 234234 --yes',
    '$ balena env rm 345345 --config',
    '$ balena env rm 456456 --service',
    '$ balena env rm 567567 --device',
    '$ balena env rm 678678 --device --config',
    '$ balena env rm 789789 --device --service --yes',
];
EnvRmCmd.args = [
    {
        name: 'id',
        required: true,
        description: "variable's numeric database ID",
        parse: (input) => validation_1.parseAsInteger(input, 'id'),
    },
];
EnvRmCmd.usage = 'env rm <id>';
EnvRmCmd.flags = {
    config: ec.booleanConfig,
    device: ec.booleanDevice,
    service: ec.booleanService,
    yes: command_1.flags.boolean({
        char: 'y',
        description: 'do not prompt for confirmation before deleting the variable',
        default: false,
    }),
};
