// resin-compose-parse/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationError = exports.validate = exports.SchemaVersion = exports.SchemaError = exports.DEFAULT_SCHEMA_VERSION = exports.parse = exports.normalize = exports.defaultComposition = void 0;
const compose_1 = require("./compose");
Object.defineProperty(exports, "defaultComposition", { enumerable: true, get: function () { return compose_1.defaultComposition; } });
Object.defineProperty(exports, "normalize", { enumerable: true, get: function () { return compose_1.normalize; } });
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return compose_1.parse; } });
const errors_1 = require("./errors");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return errors_1.ValidationError; } });
const schemas_1 = require("./schemas");
Object.defineProperty(exports, "DEFAULT_SCHEMA_VERSION", { enumerable: true, get: function () { return schemas_1.DEFAULT_SCHEMA_VERSION; } });
Object.defineProperty(exports, "SchemaError", { enumerable: true, get: function () { return schemas_1.SchemaError; } });
Object.defineProperty(exports, "SchemaVersion", { enumerable: true, get: function () { return schemas_1.SchemaVersion; } });
Object.defineProperty(exports, "validate", { enumerable: true, get: function () { return schemas_1.validate; } });
