// balena-cli/build/commands/key/add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("../../command");
const errors_1 = require("../../errors");
const cf = require("../../utils/common-flags");
const lazy_1 = require("../../utils/lazy");
class KeyAddCmd extends command_1.default {
    async run() {
        const { args: params } = this.parse(KeyAddCmd);
        let key;
        if (params.path != null) {
            const { readFile } = (await Promise.resolve().then(() => require('fs'))).promises;
            key = await readFile(params.path, 'utf8');
        }
        else if (this.stdin.length > 0) {
            key = this.stdin;
        }
        else {
            throw new errors_1.ExpectedError('No public key file or path provided.');
        }
        await lazy_1.getBalenaSdk().models.key.create(params.name, key);
    }
}
exports.default = KeyAddCmd;
KeyAddCmd.description = lazy_1.stripIndent `
		Add an SSH key to balenaCloud.

		Add an SSH key to the balenaCloud account of the logged in user.

		If \`path\` is omitted, the command will attempt to read the SSH key from stdin.

		About SSH keys  
		An "SSH key" actually consists of a public/private key pair. A typical name
		for the private key file is "id_rsa", and a typical name for the public key
		file is "id_rsa.pub". Both key files are saved to your computer (with the
		private key optionally protected by a password), but only the public key is
		saved to your balena account.  This means that if you change computers or
		otherwise lose the private key, you cannot recover the private key through
		your balena account. You can however add new keys, and delete the old ones.
		
		To generate a new SSH key pair, a nice guide can be found in GitHub's docs:
		https://help.github.com/en/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent
		Skip the step about adding the key to a GitHub account, and instead add it to
		your balena account.
	`;
KeyAddCmd.examples = [
    '$ balena key add Main ~/.ssh/id_rsa.pub',
    '$ cat ~/.ssh/id_rsa.pub | balena key add Main',
    '# Windows 10 (cmd.exe prompt) example',
    '$ balena key add Main %userprofile%.sshid_rsa.pub',
];
KeyAddCmd.args = [
    {
        name: 'name',
        description: 'the SSH key name',
        required: true,
    },
    {
        name: `path`,
        description: `the path to the public key file`,
    },
];
KeyAddCmd.usage = 'key add <name> [path]';
KeyAddCmd.flags = {
    help: cf.help,
};
KeyAddCmd.authenticated = true;
KeyAddCmd.readStdin = true;
