// balena-cli/build/commands/key/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("../../command");
const cf = require("../../utils/common-flags");
const lazy_1 = require("../../utils/lazy");
const validation_1 = require("../../utils/validation");
class KeyCmd extends command_1.default {
    async run() {
        const { args: params } = this.parse(KeyCmd);
        const key = await lazy_1.getBalenaSdk().models.key.get(params.id);
        const displayKey = {
            id: key.id,
            name: key.title,
        };
        console.log(lazy_1.getVisuals().table.vertical(displayKey, ['id', 'name']));
        console.log('\n' + key.public_key);
    }
}
exports.default = KeyCmd;
KeyCmd.description = lazy_1.stripIndent `
		Display an SSH key.

		Display a single SSH key registered in balenaCloud for the logged in user.
	`;
KeyCmd.examples = ['$ balena key 17'];
KeyCmd.args = [
    {
        name: 'id',
        description: 'balenaCloud ID for the SSH key',
        parse: (x) => validation_1.parseAsInteger(x, 'id'),
        required: true,
    },
];
KeyCmd.usage = 'key <id>';
KeyCmd.flags = {
    help: cf.help,
};
KeyCmd.authenticated = true;
