// balena-release/build/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConflictError = exports.HttpResponseError = exports.ServerError = exports.BadRequestError = exports.UnauthorisedError = exports.UniqueConstraintError = exports.ObjectDoesNotExistError = void 0;
const typed_error_1 = require("typed-error");
class ObjectDoesNotExistError extends typed_error_1.TypedError {
}
exports.ObjectDoesNotExistError = ObjectDoesNotExistError;
class UniqueConstraintError extends typed_error_1.TypedError {
}
exports.UniqueConstraintError = UniqueConstraintError;
class UnauthorisedError extends typed_error_1.TypedError {
}
exports.UnauthorisedError = UnauthorisedError;
class BadRequestError extends typed_error_1.TypedError {
}
exports.BadRequestError = BadRequestError;
class ServerError extends typed_error_1.TypedError {
}
exports.ServerError = ServerError;
class HttpResponseError extends typed_error_1.TypedError {
    constructor(message, statusCode) {
        super(message);
        this.statusCode = statusCode;
    }
}
exports.HttpResponseError = HttpResponseError;
class ConflictError extends HttpResponseError {
}
exports.ConflictError = ConflictError;
