// balena-sdk/es2018/util/device-actions/device-actions-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceActionsService = void 0;
const DEVICE_ACTIONS_API_VERSION = 'v1';
class DeviceActionsService {
    constructor(deviceUrlsBase, request) {
        this.request = request;
        this.startAction = ({ uuid, actionName, params, extraOptions, }) => this.makeActionRequest({
            method: 'POST',
            uuid,
            actionNameOrId: actionName,
            params,
            extraOptions,
        });
        this.getActionStatus = ({ uuid, actionId, extraOptions, }) => this.makeActionRequest({
            method: 'GET',
            uuid,
            actionNameOrId: actionId,
            extraOptions,
        });
        this.makeActionRequest = async ({ method, uuid, actionNameOrId, params, extraOptions, }) => {
            const data = params ? { parameters: params } : null;
            const { body } = await this.request.send({
                method,
                url: `${this.actionsEndpoint}/${uuid}/${actionNameOrId}`,
                body: data,
                ...extraOptions,
            });
            return body;
        };
        this.actionsEndpoint = `https://actions.${deviceUrlsBase}/${DEVICE_ACTIONS_API_VERSION}`;
    }
}
exports.DeviceActionsService = DeviceActionsService;
