// file-disk/build/fs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.write = exports.read = exports.fdatasync = exports.fstat = exports.close = exports.open = void 0;
const Bluebird = require("bluebird");
const fs = require("fs");
exports.open = Bluebird.promisify(fs.open, { context: fs });
exports.close = Bluebird.promisify(fs.close, { context: fs });
exports.fstat = Bluebird.promisify(fs.fstat, { context: fs });
exports.fdatasync = Bluebird.promisify(fs.fdatasync, { context: fs });
exports.read = (fd, buf, offset, length, position) => {
    return new Promise((resolve, reject) => {
        fs.read(fd, buf, offset, length, position, (error, bytesRead, buffer) => {
            if (error !== null) {
                reject(error);
            }
            else {
                resolve({ bytesRead, buffer });
            }
        });
    });
};
exports.write = (fd, buf, offset, length, position) => {
    return new Promise((resolve, reject) => {
        fs.write(fd, buf, offset, length, position, (error, bytesWritten, buffer) => {
            if (error !== null) {
                reject(error);
            }
            else {
                resolve({ bytesWritten, buffer });
            }
        });
    });
};
