// resin-bundle-resolve/build/resolvers/dockerfile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class DockerfileResolver {
    constructor() {
        this.priority = 0;
        this.name = 'Standard Dockerfile';
        this.gotDockerfile = false;
    }
    entry(file) {
        this.gotDockerfile = true;
        this.dockerfileContents = file.contents.toString();
    }
    needsEntry(entryPath, specifiedDockerfilePath) {
        return entryPath.unparsed === (specifiedDockerfilePath || 'Dockerfile');
    }
    isSatisfied() {
        return this.gotDockerfile;
    }
    resolve(_bundle, specifiedDockerfilePath) {
        return Promise.resolve({
            name: specifiedDockerfilePath || 'Dockerfile',
            contents: Buffer.from(this.dockerfileContents),
            size: this.dockerfileContents.length,
        });
    }
    getDockerfileContents() {
        return this.dockerfileContents;
    }
    getCanonicalName(filename) {
        return filename;
    }
}
exports.DockerfileResolver = DockerfileResolver;
exports.default = DockerfileResolver;
