// resin-bundle-resolve/build/resolvers/nodeResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Bluebird = require("bluebird");
const _ = require("lodash");
const request = require("request");
const semver = require("semver");
const getAsync = Bluebird.promisify(request.get);
const BluebirdLRU = require("bluebird-lru-cache");
const versionTest = RegExp.prototype.test.bind(/^[0-9]+\.[0-9]+\.[0-9]+$/);
const versionCache = new BluebirdLRU({
    maxAge: 3600 * 1000,
    fetchFn: (deviceType) => {
        const get = (prev, url) => {
            return Promise.resolve(getAsync({
                url,
                json: true,
            })
                .get('body')
                .then((res) => {
                const curr = _(res.results)
                    .map('name')
                    .filter(versionTest)
                    .value();
                const tags = prev.concat(curr);
                if (res.next != null) {
                    return get(tags, res.next);
                }
                else {
                    return tags;
                }
            }));
        };
        return get([], `https://hub.docker.com/v2/repositories/resin/${deviceType}-node/tags/?page_size=100`);
    },
});
class NodeResolver {
    constructor() {
        this.priority = 0;
        this.name = 'NodeJS';
        this.hasScripts = false;
    }
    entry(file) {
        if (file.name === 'package.json') {
            this.packageJsonContent = file.contents;
        }
        else if (file.name === 'wscript' || _.endsWith(file.name, '.gyp')) {
            this.hasScripts = true;
        }
    }
    needsEntry(entryPath, specifiedDockerfilePath) {
        const unparsed = entryPath.unparsed;
        return (!specifiedDockerfilePath &&
            (unparsed === 'package.json' ||
                unparsed === 'wscript' ||
                (!unparsed.includes('/') && entryPath.ext === '.gyp')));
    }
    isSatisfied(_bundle) {
        return this.packageJsonContent != null;
    }
    resolve(bundle, _specifiedDockerfilePath) {
        return Promise.resolve(Bluebird.try(() => JSON.parse(this.packageJsonContent.toString()))
            .catch((e) => {
            throw new Error(`package.json: ${e.message}`);
        })
            .then(packageJson => {
            if (typeof packageJson !== 'object') {
                throw new Error('package.json: must be a JSON object');
            }
            this.hasScripts =
                this.hasScripts ||
                    _(packageJson.scripts)
                        .pick('preinstall', 'install', 'postinstall')
                        .size() > 0;
            const nodeEngine = _.get(packageJson, 'engines.node');
            if (nodeEngine == null) {
                throw new Error('package.json: engines.node must be specified');
            }
            if (!_.isString(nodeEngine)) {
                throw new Error('package.json: engines.node must be a string if present');
            }
            const range = nodeEngine;
            return versionCache.get(bundle.deviceType).then(versions => {
                const nodeVersion = semver.maxSatisfying(versions, range);
                if (nodeVersion == null) {
                    throw new Error(`Couldn't satisfy node version ${range}`);
                }
                let dockerfile;
                if (this.hasScripts) {
                    dockerfile = `
						FROM resin/${bundle.deviceType}-node:${nodeVersion}
						RUN mkdir -p /usr/src/app && ln -s /usr/src/app /app
						WORKDIR /usr/src/app
						COPY . /usr/src/app
						RUN DEBIAN_FRONTEND=noninteractive JOBS=MAX npm install --unsafe-perm
						CMD [ "npm", "start" ]
						`;
                }
                else {
                    dockerfile = `
						FROM resin/${bundle.deviceType}-node:${nodeVersion}-onbuild
						RUN ln -s /usr/src/app /app
					`;
                }
                this.dockerfileContents = dockerfile;
                return {
                    name: 'Dockerfile',
                    size: dockerfile.length,
                    contents: Buffer.from(dockerfile),
                };
            });
        }));
    }
    getCanonicalName(_specifiedPath) {
        throw new Error('getCanonicalName called on unsupported resolver NodeJS');
    }
}
exports.NodeResolver = NodeResolver;
exports.default = NodeResolver;
