// balena-cli/build/commands/app/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("../../command");
const cf = require("../../utils/common-flags");
const ca = require("../../utils/common-args");
const lazy_1 = require("../../utils/lazy");
const messages_1 = require("../../utils/messages");
class AppCmd extends command_1.default {
    async run() {
        var _a;
        const { args: params } = this.parse(AppCmd);
        const { getApplication } = await Promise.resolve().then(() => require('../../utils/sdk'));
        const application = (await getApplication(lazy_1.getBalenaSdk(), params.application, {
            $expand: {
                is_for__device_type: { $select: 'slug' },
                should_be_running__release: { $select: 'commit' },
            },
        }));
        application.device_type = application.is_for__device_type[0].slug;
        application.commit = (_a = application.should_be_running__release[0]) === null || _a === void 0 ? void 0 : _a.commit;
        console.log(`== ${application.app_name}`);
        console.log(lazy_1.getVisuals().table.vertical(application, [
            'id',
            'device_type',
            'slug',
            'commit',
        ]));
    }
}
exports.default = AppCmd;
AppCmd.description = lazy_1.stripIndent `
		Display information about a single application.

		Display detailed information about a single balena application.

		${messages_1.applicationIdInfo.split('\n').join('\n\t\t')}
`;
AppCmd.examples = ['$ balena app MyApp', '$ balena app myorg/myapp'];
AppCmd.args = [ca.applicationRequired];
AppCmd.usage = 'app <nameOrSlug>';
AppCmd.flags = {
    help: cf.help,
};
AppCmd.authenticated = true;
AppCmd.primary = true;
