// balena-cli/build/commands/config/inject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const command_2 = require("../../command");
const cf = require("../../utils/common-flags");
const lazy_1 = require("../../utils/lazy");
class ConfigInjectCmd extends command_2.default {
    async run() {
        const { args: params, flags: options } = this.parse(ConfigInjectCmd);
        const { promisify } = await Promise.resolve().then(() => require('util'));
        const umountAsync = promisify((await Promise.resolve().then(() => require('umount'))).umount);
        const drive = options.drive || (await lazy_1.getVisuals().drive('Select the device/OS drive'));
        await umountAsync(drive);
        const fs = await Promise.resolve().then(() => require('fs'));
        const configJSON = JSON.parse(await fs.promises.readFile(params.file, 'utf8'));
        const config = await Promise.resolve().then(() => require('balena-config-json'));
        await config.write(drive, options.type, configJSON);
        console.info('Done');
    }
}
exports.default = ConfigInjectCmd;
ConfigInjectCmd.description = lazy_1.stripIndent `
		Inject a configuration file into a device or OS image.

		Inject a config.json file to the mounted filesystem,
		e.g. the SD card of a provisioned device or balenaOS image.
	`;
ConfigInjectCmd.examples = [
    '$ balena config inject my/config.json --type raspberrypi3',
    '$ balena config inject my/config.json --type raspberrypi3 --drive /dev/disk2',
];
ConfigInjectCmd.args = [
    {
        name: 'file',
        description: 'the path to the config.json file to inject',
        required: true,
    },
];
ConfigInjectCmd.usage = 'config inject <file>';
ConfigInjectCmd.flags = {
    type: command_1.flags.string({
        description: 'device type (Check available types with `balena devices supported`)',
        char: 't',
        required: true,
    }),
    drive: command_1.flags.string({
        description: 'device filesystem or OS image location',
        char: 'd',
    }),
    help: cf.help,
};
ConfigInjectCmd.authenticated = true;
ConfigInjectCmd.root = true;
