// resin-multibuild/build/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateBuildTasks = void 0;
const _ = require("lodash");
const tar = require("tar-stream");
function generateBuildTasks(images, buildMetadata) {
    return _.map(images, img => {
        if (_.isString(img.image)) {
            return {
                external: true,
                imageName: img.image,
                serviceName: img.serviceName,
                resolved: false,
                buildMetadata,
            };
        }
        else {
            if (img.image.context == null) {
                throw new Error('Must have a context specified with a Dockerfile');
            }
            return _.merge({
                external: false,
                serviceName: img.serviceName,
                buildStream: tar.pack(),
                resolved: false,
                buildMetadata,
            }, img.image.dockerfile != null
                ? { dockerfilePath: img.image.dockerfile }
                : {}, _.omit(img.image, 'image'));
        }
    });
}
exports.generateBuildTasks = generateBuildTasks;
