// balena-sdk/es2018/auth.js
"use strict";
/*
Copyright 2016 Balena

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const errors = require("balena-errors");
const memoizee = require("memoizee");
const getAuth = function (deps, opts) {
    const { auth: authBase, pubsub, request } = deps;
    const { apiUrl } = opts;
    const normalizeAuthError = function (err) {
        if (err.statusCode === 401) {
            return new errors.BalenaNotLoggedIn();
        }
        else if (err.code === 'BalenaMalformedToken') {
            return new errors.BalenaNotLoggedIn();
        }
        else {
            return err;
        }
    };
    const wrapAuthFn = (eventName, fn) => async function () {
        try {
            return await fn.apply(authBase, arguments);
        }
        finally {
            pubsub.publish(eventName);
        }
    };
    const auth = {
        ...authBase,
        setKey: wrapAuthFn('auth.keyChange', authBase.setKey),
        removeKey: wrapAuthFn('auth.keyChange', authBase.removeKey),
    };
    /**
     * @namespace balena.auth.twoFactor
     * @memberof balena.auth
     */
    const twoFactor = require('./2fa').default({
        ...deps,
        auth,
    }, opts);
    const userWhoami = async () => {
        const { body } = await request.send({
            method: 'GET',
            url: '/user/v1/whoami',
            baseUrl: apiUrl,
        });
        return body;
    };
    const memoizedUserWhoami = memoizee(userWhoami, {
        primitive: true,
        promise: true,
    });
    const getUserDetails = async (noCache = false) => {
        if (noCache) {
            memoizedUserWhoami.clear();
        }
        try {
            return await memoizedUserWhoami();
        }
        catch (err) {
            throw normalizeAuthError(err);
        }
    };
    /**
     * @summary Return current logged in username
     * @name whoami
     * @public
     * @function
     * @memberof balena.auth
     *
     * @description This will only work if you used {@link balena.auth.login} to log in.
     *
     * @fulfil {(String|undefined)} - username, if it exists
     * @returns {Promise}
     *
     * @example
     * balena.auth.whoami().then(function(username) {
     * 	if (!username) {
     * 		console.log('I\'m not logged in!');
     * 	} else {
     * 		console.log('My username is:', username);
     * 	}
     * });
     *
     * @example
     * balena.auth.whoami(function(error, username) {
     * 	if (error) throw error;
     *
     * 	if (!username) {
     * 		console.log('I\'m not logged in!');
     * 	} else {
     * 		console.log('My username is:', username);
     * 	}
     * });
     */
    async function whoami() {
        try {
            const userDetails = await getUserDetails();
            return userDetails === null || userDetails === void 0 ? void 0 : userDetails.username;
        }
        catch (err) {
            if (err instanceof errors.BalenaNotLoggedIn) {
                return;
            }
            throw err;
        }
    }
    /**
     * @summary Authenticate with the server
     * @name authenticate
     * @protected
     * @function
     * @memberof balena.auth
     *
     * @description You should use {@link balena.auth.login} when possible,
     * as it takes care of saving the token and email as well.
     *
     * Notice that if `credentials` contains extra keys, they'll be discarted
     * by the server automatically.
     *
     * @param {Object} credentials - in the form of email, password
     * @param {String} credentials.email - the email
     * @param {String} credentials.password - the password
     *
     * @fulfil {String} - session token
     * @returns {Promise}
     *
     * @example
     * balena.auth.authenticate(credentials).then(function(token) {
     * 	console.log('My token is:', token);
     * });
     *
     * @example
     * balena.auth.authenticate(credentials, function(error, token) {
     * 	if (error) throw error;
     * 	console.log('My token is:', token);
     * });
     */
    async function authenticate(credentials) {
        try {
            const { body } = await request.send({
                method: 'POST',
                baseUrl: apiUrl,
                url: '/login_',
                body: {
                    username: credentials.email,
                    password: String(credentials.password),
                },
                sendToken: false,
            });
            return body;
        }
        catch (err) {
            if (err.statusCode === 401) {
                throw new errors.BalenaInvalidLoginCredentials();
            }
            if (err.statusCode === 429) {
                throw new errors.BalenaTooManyRequests();
            }
            throw err;
        }
    }
    /**
     * @summary Login
     * @name login
     * @public
     * @function
     * @memberof balena.auth
     *
     * @description If the login is successful, the token is persisted between sessions.
     *
     * @param {Object} credentials - in the form of email, password
     * @param {String} credentials.email - the email
     * @param {String} credentials.password - the password
     *
     * @returns {Promise}
     *
     * @example
     * balena.auth.login(credentials);
     *
     * @example
     * balena.auth.login(credentials, function(error) {
     * 	if (error) throw error;
     * });
     */
    async function login(credentials) {
        memoizedUserWhoami.clear();
        const token = await authenticate(credentials);
        await auth.setKey(token);
    }
    /**
     * @summary Login with a token or api key
     * @name loginWithToken
     * @public
     * @function
     * @memberof balena.auth
     *
     * @description Login to balena with a session token or api key instead of with credentials.
     *
     * @param {String} authToken - the auth token
     * @returns {Promise}
     *
     * @example
     * balena.auth.loginWithToken(authToken);
     *
     * @example
     * balena.auth.loginWithToken(authToken, function(error) {
     * 	if (error) throw error;
     * });
     */
    function loginWithToken(authToken) {
        memoizedUserWhoami.clear();
        return auth.setKey(authToken);
    }
    /**
     * @summary Check if you're logged in
     * @name isLoggedIn
     * @public
     * @function
     * @memberof balena.auth
     *
     * @fulfil {Boolean} - is logged in
     * @returns {Promise}
     *
     * @example
     * balena.auth.isLoggedIn().then(function(isLoggedIn) {
     * 	if (isLoggedIn) {
     * 		console.log('I\'m in!');
     * 	} else {
     * 		console.log('Too bad!');
     * 	}
     * });
     *
     * @example
     * balena.auth.isLoggedIn(function(error, isLoggedIn) {
     * 	if (error) throw error;
     *
     * 	if (isLoggedIn) {
     * 		console.log('I\'m in!');
     * 	} else {
     * 		console.log('Too bad!');
     * 	}
     * });
     */
    async function isLoggedIn() {
        try {
            await getUserDetails(true);
            return true;
        }
        catch (err) {
            if (err instanceof errors.BalenaNotLoggedIn) {
                return false;
            }
            throw err;
        }
    }
    /**
     * @summary Get current logged in user's raw API key or session token
     * @name getToken
     * @public
     * @function
     * @memberof balena.auth
     *
     * @description This will only work if you used {@link balena.auth.login} to log in.
     *
     * @fulfil {String} - raw API key or session token
     * @returns {Promise}
     *
     * @example
     * balena.auth.getToken().then(function(token) {
     * 	console.log(token);
     * });
     *
     * @example
     * balena.auth.getToken(function(error, token) {
     * 	if (error) throw error;
     * 	console.log(token);
     * });
     */
    function getToken() {
        return auth.getKey().catch(function (err) {
            throw normalizeAuthError(err);
        });
    }
    /**
     * @summary Get current logged in user's id
     * @name getUserId
     * @public
     * @function
     * @memberof balena.auth
     *
     * @description This will only work if you used {@link balena.auth.login} to log in.
     *
     * @fulfil {Number} - user id
     * @returns {Promise}
     *
     * @example
     * balena.auth.getUserId().then(function(userId) {
     * 	console.log(userId);
     * });
     *
     * @example
     * balena.auth.getUserId(function(error, userId) {
     * 	if (error) throw error;
     * 	console.log(userId);
     * });
     */
    async function getUserId() {
        const { id } = await getUserDetails();
        return id;
    }
    /**
     * @summary Get current logged in user's email
     * @name getEmail
     * @public
     * @function
     * @memberof balena.auth
     *
     * @description This will only work if you used {@link balena.auth.login} to log in.
     *
     * @fulfil {String} - user email
     * @returns {Promise}
     *
     * @example
     * balena.auth.getEmail().then(function(email) {
     * 	console.log(email);
     * });
     *
     * @example
     * balena.auth.getEmail(function(error, email) {
     * 	if (error) throw error;
     * 	console.log(email);
     * });
     */
    async function getEmail() {
        const { email } = await getUserDetails();
        return email;
    }
    /**
     * @summary Logout
     * @name logout
     * @public
     * @function
     * @memberof balena.auth
     *
     * @returns {Promise}
     *
     * @example
     * balena.auth.logout();
     *
     * @example
     * balena.auth.logout(function(error) {
     * 	if (error) throw error;
     * });
     */
    function logout() {
        memoizedUserWhoami.clear();
        return auth.removeKey();
    }
    /**
     * @summary Register a user account
     * @name register
     * @public
     * @function
     * @memberof balena.auth
     *
     * @param {Object} credentials - in the form of username, password and email
     * @param {String} credentials.email - the email
     * @param {String} credentials.password - the password
     * @param {(String|undefined)} [credentials.'g-recaptcha-response'] - the captcha response
     *
     * @fulfil {String} - session token
     * @returns {Promise}
     *
     * @example
     * balena.auth.register({
     * 	email: 'johndoe@gmail.com',
     * 	password: 'secret'
     * }).then(function(token) {
     * 	console.log(token);
     * });
     *
     * @example
     * balena.auth.register({
     * 	email: 'johndoe@gmail.com',
     * 	password: 'secret'
     * }, function(error, token) {
     * 	if (error) throw error;
     * 	console.log(token);
     * });
     */
    async function register(credentials) {
        const { body } = await request.send({
            method: 'POST',
            url: '/user/register',
            baseUrl: apiUrl,
            body: credentials,
            sendToken: false,
        });
        return body;
    }
    return {
        twoFactor: require('./util/callbacks').addCallbackSupportToModule(twoFactor),
        whoami,
        authenticate,
        login,
        loginWithToken,
        isLoggedIn,
        getToken,
        getUserId,
        getEmail,
        logout,
        register,
    };
};
exports.default = getAuth;
