// balena-cli/build/commands/device/identify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("../../command");
const cf = require("../../utils/common-flags");
const lazy_1 = require("../../utils/lazy");
const validation_1 = require("../../utils/validation");
const errors_1 = require("../../errors");
class DeviceIdentifyCmd extends command_1.default {
    async run() {
        var _a;
        const { args: params } = this.parse(DeviceIdentifyCmd);
        const balena = lazy_1.getBalenaSdk();
        try {
            await balena.models.device.identify(params.uuid);
        }
        catch (e) {
            if ((_a = e.message) === null || _a === void 0 ? void 0 : _a.toLowerCase().includes('online')) {
                throw new errors_1.ExpectedError(`Device ${params.uuid} is not online`);
            }
            else {
                throw e;
            }
        }
    }
}
exports.default = DeviceIdentifyCmd;
DeviceIdentifyCmd.description = lazy_1.stripIndent `
		Identify a device.

		Identify a device by making the ACT LED blink (Raspberry Pi).
		`;
DeviceIdentifyCmd.examples = ['$ balena device identify 23c73a1'];
DeviceIdentifyCmd.args = [
    {
        name: 'uuid',
        description: 'the uuid of the device to identify',
        parse: (dev) => validation_1.tryAsInteger(dev),
        required: true,
    },
];
DeviceIdentifyCmd.usage = 'device identify <uuid>';
DeviceIdentifyCmd.flags = {
    help: cf.help,
};
DeviceIdentifyCmd.authenticated = true;
