// balena-cli/build/commands/device/reboot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("../../command");
const cf = require("../../utils/common-flags");
const lazy_1 = require("../../utils/lazy");
const validation_1 = require("../../utils/validation");
class DeviceRebootCmd extends command_1.default {
    async run() {
        const { args: params, flags: options } = this.parse(DeviceRebootCmd);
        const balena = lazy_1.getBalenaSdk();
        await balena.models.device.reboot(params.uuid, options);
    }
}
exports.default = DeviceRebootCmd;
DeviceRebootCmd.description = lazy_1.stripIndent `
		Restart a device.

		Remotely reboot a device.
		`;
DeviceRebootCmd.examples = ['$ balena device reboot 23c73a1'];
DeviceRebootCmd.args = [
    {
        name: 'uuid',
        description: 'the uuid of the device to reboot',
        parse: (dev) => validation_1.tryAsInteger(dev),
        required: true,
    },
];
DeviceRebootCmd.usage = 'device reboot <uuid>';
DeviceRebootCmd.flags = {
    force: cf.force,
    help: cf.help,
};
DeviceRebootCmd.authenticated = true;
