// balena-cli/build/commands/tag/set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("../../command");
const cf = require("../../utils/common-flags");
const lazy_1 = require("../../utils/lazy");
const messages_1 = require("../../utils/messages");
class TagSetCmd extends command_1.default {
    async run() {
        var _a;
        const { args: params, flags: options } = this.parse(TagSetCmd);
        options.application = options.application || options.app;
        delete options.app;
        const balena = lazy_1.getBalenaSdk();
        if (!options.application && !options.device && !options.release) {
            const { ExpectedError } = await Promise.resolve().then(() => require('../../errors'));
            throw new ExpectedError(TagSetCmd.missingResourceMessage);
        }
        (_a = params.value) !== null && _a !== void 0 ? _a : (params.value = '');
        const { tryAsInteger } = await Promise.resolve().then(() => require('../../utils/validation'));
        if (options.application) {
            const { getTypedApplicationIdentifier } = await Promise.resolve().then(() => require('../../utils/sdk'));
            return balena.models.application.tags.set(await getTypedApplicationIdentifier(balena, options.application), params.tagKey, params.value);
        }
        if (options.device) {
            return balena.models.device.tags.set(tryAsInteger(options.device), params.tagKey, params.value);
        }
        if (options.release) {
            const { disambiguateReleaseParam } = await Promise.resolve().then(() => require('../../utils/normalization'));
            const releaseParam = await disambiguateReleaseParam(balena, options.release);
            return balena.models.release.tags.set(releaseParam, params.tagKey, params.value);
        }
    }
}
exports.default = TagSetCmd;
TagSetCmd.description = lazy_1.stripIndent `
		Set a tag on an application, device or release.

		Set a tag on an application, device or release.

		You can optionally provide a value to be associated with the created
		tag, as an extra argument after the tag key. If a value isn't
		provided, a tag with an empty value is created.

		${messages_1.applicationIdInfo.split('\n').join('\n\t\t')}
	`;
TagSetCmd.examples = [
    '$ balena tag set mySimpleTag --application MyApp',
    '$ balena tag set mySimpleTag -a myorg/myapp',
    '$ balena tag set myCompositeTag myTagValue --application MyApp',
    '$ balena tag set myCompositeTag myTagValue --device 7cf02a6',
    '$ balena tag set myCompositeTag "my tag value with whitespaces" --device 7cf02a6',
    '$ balena tag set myCompositeTag myTagValue --release 1234',
    '$ balena tag set myCompositeTag --release 1234',
    '$ balena tag set myCompositeTag --release b376b0e544e9429483b656490e5b9443b4349bd6',
];
TagSetCmd.args = [
    {
        name: 'tagKey',
        description: 'the key string of the tag',
        required: true,
    },
    {
        name: 'value',
        description: 'the optional value associated with the tag',
        required: false,
    },
];
TagSetCmd.usage = 'tag set <tagKey> [value]';
TagSetCmd.flags = {
    application: {
        ...cf.application,
        exclusive: ['app', 'device', 'release'],
    },
    app: {
        ...cf.app,
        exclusive: ['application', 'device', 'release'],
    },
    device: {
        ...cf.device,
        exclusive: ['app', 'application', 'release'],
    },
    release: {
        ...cf.release,
        exclusive: ['app', 'application', 'device'],
    },
    help: cf.help,
};
TagSetCmd.authenticated = true;
TagSetCmd.missingResourceMessage = lazy_1.stripIndent `
					To set a resource tag, you must provide exactly one of:

					  * An application, with --application <appNameOrSlug>
					  * A device, with --device <uuid>
					  * A release, with --release <id or commit>

					See the help page for examples:

					  $ balena help tag set
	`;
