// balena-cli/build/utils/device/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceUnavailableAPIError = exports.BadRequestDeviceAPIError = exports.DeviceAPIError = exports.BuildError = void 0;
const _ = require("lodash");
const errors_1 = require("../../errors");
class BuildError extends errors_1.ExpectedError {
    constructor(failures) {
        super('Build error');
        this.failures = failures;
    }
    toString() {
        let str = 'Some services failed to build:\n';
        _.each(this.failures, (failure) => {
            str += `\t${failure.serviceName}: ${failure.error.message}\n`;
        });
        return str;
    }
    getServiceError(serviceName) {
        const failure = _.find(this.failures, (f) => f.serviceName === serviceName);
        if (failure == null) {
            return 'Unknown build failure';
        }
        return failure.error.message;
    }
}
exports.BuildError = BuildError;
class DeviceAPIError extends errors_1.ExpectedError {
}
exports.DeviceAPIError = DeviceAPIError;
class BadRequestDeviceAPIError extends DeviceAPIError {
}
exports.BadRequestDeviceAPIError = BadRequestDeviceAPIError;
class ServiceUnavailableAPIError extends DeviceAPIError {
}
exports.ServiceUnavailableAPIError = ServiceUnavailableAPIError;
