// livepush/build/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidArgumentError = exports.RuntimeError = exports.InternalInconsistencyError = exports.ContainerNotRunningError = exports.UnsupportedError = exports.DockerfileParseError = void 0;
const typed_error_1 = require("typed-error");
class DockerfileParseError extends typed_error_1.TypedError {
}
exports.DockerfileParseError = DockerfileParseError;
class UnsupportedError extends typed_error_1.TypedError {
}
exports.UnsupportedError = UnsupportedError;
class ContainerNotRunningError extends typed_error_1.TypedError {
}
exports.ContainerNotRunningError = ContainerNotRunningError;
class InternalInconsistencyError extends typed_error_1.TypedError {
}
exports.InternalInconsistencyError = InternalInconsistencyError;
class RuntimeError extends typed_error_1.TypedError {
}
exports.RuntimeError = RuntimeError;
class InvalidArgumentError extends typed_error_1.TypedError {
}
exports.InvalidArgumentError = InvalidArgumentError;
