// balena-pine/es2018/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BalenaPine = void 0;
const url = require("url");
const errors = require("balena-errors");
const pinejs_client_core_1 = require("pinejs-client-core");
class BalenaPine extends pinejs_client_core_1.PinejsClientCore {
    constructor(params, backendParams) {
        super({
            ...params,
            apiPrefix: url.resolve(backendParams.apiUrl, `/${backendParams.apiVersion}/`),
        });
        this.backendParams = backendParams;
        this.backendParams = backendParams;
        this.API_URL = backendParams.apiUrl;
        this.API_VERSION = backendParams.apiVersion;
    }
    async _request(options) {
        const { apiKey, apiUrl, auth, request } = this.backendParams;
        const hasKey = await auth.hasKey();
        const authenticated = hasKey || (apiKey != null && apiKey.length > 0);
        options = {
            apiKey,
            baseUrl: apiUrl,
            sendToken: authenticated && !options.anonymous,
            ...options,
        };
        try {
            const { body } = await request.send(options);
            return body;
        }
        catch (err) {
            if (err.statusCode !== 401) {
                throw err;
            }
            if (options.anonymous) {
                throw err;
            }
            if (!authenticated) {
                throw new errors.BalenaNotLoggedIn();
            }
            throw err;
        }
    }
}
exports.BalenaPine = BalenaPine;
